### DISCUSSION TITLE
SSO: Oauth2 providers does not return code and state on Mobile App

### DISCUSSION THREAD

**Discussion by codihaus** [directus/directus#16741] [Feature Requests]
Short description: As a mobile (and web app) developer, I would like to login by accessToken from Facebook after authorized via Facebook SDK

Real usage case: let's say we already have Directus with OAuth2 Facebook configured

- On Mobile Application, create the button to open Directus admin by webview
- Click on Facebook login
- Error (please see the attached error below)

Reason: Facebook stopped allowing single sign on via webview. Link: https://developers.facebook.com/docs/facebook-login/android/deprecating-webviews

===========

Current flow of login of Directus is done via Web browser, implemented with "Manually flow" from https://developers.facebook.com/docs/facebook-login/guides/advanced/manual-flow, it will only work for Desktop or Open by Browser. Using web view will not work.

So, in this case, we must use Facebook JS SDK to login, but with this SDK, it only allow you to init the process without code & parameter. (Link SDK: https://www.npmjs.com/package/react-native-fbsdk-next)

===========

My suggestion: I would suggest to allow user login by accessToken only, as mentioned above, with SDK, it only return accessToken after user accept login.

**Comment by codihaus** [comment]
Here's the error when you open link via Webview:

![image](https://user-images.githubusercontent.com/68498487/206216033-39b16442-d6fd-4088-acae-d4b2f2e24513.png)

**Reply by codihaus** [reply]
For more inforamation from AppleID. Link: https://developer.apple.com/news/?id=edwud51q

Statement: "New apps containing these frameworks are no longer accepted by the App Store"

That mean, we must use this service https://developer.apple.com/documentation/authenticationservices, which only return accessToken only, as well.

**Reply by codihaus** [reply]
More information: https://developer.apple.com/forums/thread/101396

Hacker News got reject by using webview as well.

**Reply by codihaus** [reply]
With the current setup, if we must follow:

- We use Directus Admin (1)
- Click to open web browser (not webview) and authorize(2)

(2) will lead to app rejected by Apple App Store, as information above, I think it caused by the cross-domain policy, where our button is Login with Facebook, but open 3rd party web (not facebook) and redirect users through unknown/unpredictable target for taking information.

**Comment by codihaus** [comment]
Here's my very first implementation for enhance the login Flow.

![image](https://user-images.githubusercontent.com/68498487/206215809-e9cead42-7df5-41da-8f99-00e49b815bb3.png)

With this, I can create a custom endpoint to allow login via accessToken.

**Comment by codihaus** [comment]
In term of UX:

By allowing login with accessToken only, for Mobile App & Web app:

We do not need to redirect user to anywhere, just click login with Facebook > get accessToken > post to Directus API > login.

It would be less fraction to have user use app/web app.

**Reply by codihaus** [reply]
In term of Security:

We still double check accessToken that Frontend post to API with Facebook, before grant the access. So I think it would work.

- To have someone obtain the accessToken to post to our Directus API, he must have the access to victim's Facebook Account, but if he got this access, with code - states or without, he would still be able to get the accessToken
- We also check the origin of the request, where it would come from a specific app ID, which could be encoded, to make sure it's correct.

**Reply by NilsBaumgartner1994** [reply]
I think there was already a discussion about this. The solution was to create a custom endpoint in directus, where clicked on it, it would redirect a mobile app user to the corresponding SSO Provider and then back to directus with the access_token in the URL. Then the directus custom endpoint redirects the user back to the mobile app.

**Reply by NilsBaumgartner1994** [reply]
Have a look here about the discussion: https://github.com/directus/directus/discussions/8867#discussioncomment-3742096

**Reply by codihaus** [reply]
Thank you for your reply.

My team already did it, but there' a problem.

With that flow, the application could be rejected by App store, as mentioned above, since you let user click on Facebook Login, but redirect to your domain, so it could be considered as clickbait on App store point of view.

I did show some search about this and posted on the replies above.

**Reply by NilsBaumgartner1994** [reply]
I know what you mean, but it is only used to create the nonce in the SSO flow in directus. Since the custom endpoint does not render anything and only forwards with a 301, i would think that be ok.

**Reply by TranQuangMinh** [reply]
Hi @NilsBaumgartner1994 , i have a question, 
MyApp Android open directus to login SSO and then how to re-open app via deeplink and send directus access token or refresh token to app ?

**Reply by codihaus** [reply]
> I know what you mean, but it is only used to create the nonce in the SSO flow in directus. Since the custom endpoint does not render anything and only forwards with a 301, i would think that be ok.

I see your point, how do you think about UX, user would rather stay at the application than open browser and redirect?

**Reply by NilsBaumgartner1994** [reply]
> > I know what you mean, but it is only used to create the nonce in the SSO flow in directus. Since the custom endpoint does not render anything and only forwards with a 301, i would think that be ok.
> 
> I see your point, how do you think about UX, user would rather stay at the application than open browser and redirect?

Oh I would love it to see my real native browser opening and going out of the app. I don’t really trust in app browsers, since it could be mocked. Also a mocked Browser is not allowed, but who knows. So I would prefer in terms of safety to be moved out of my app.

**Reply by codihaus** [reply]
> > > I know what you mean, but it is only used to create the nonce in the SSO flow in directus. Since the custom endpoint does not render anything and only forwards with a 301, i would think that be ok.
> > 
> > 
> > I see your point, how do you think about UX, user would rather stay at the application than open browser and redirect?
> 
> Oh I would love it to see my real native browser opening and going out of the app. I don’t really trust in app browsers, since it could be mocked. Also a mocked Browser is not allowed, but who knows. So I would prefer in terms of safety to be moved out of my app.

For "in app browser", we're not going to use in app webview, we will use their SDK to login right on app, not open any web view.

I see, thank you for your input, I will try with your way, hopefully it won't be rejected :D :D :D 

Also, how about redirect user back to the app? Do you have any experience on this?

**Reply by NilsBaumgartner1994** [reply]
> Hi @NilsBaumgartner1994 , i have a question,
> MyApp Android open directus to login SSO and then how to re-open app via deeplink and send directus access token or refresh token to app ?

Oh that’s easy :-) you create a custom endpoint in Directus with a „redirectTo“ param. Since the SSO will redirect you to your given application Directus sends you back to your app. Have an example

Endpoint to redirect with access token: Directus/curstomEndpoint/SSO/facebook?redirectTo=MobileApp&access_token=

Directus SSO Endpoint: Directus/login/sso?rddirectTo<Redirect>

In this case we set the redirect route Directus normale would set to itself, to our custom endpoint. Our custom endpoint then gets the access token and sends us back to our app. I can share a link to a gist. Meanwhile check out my added reply post at the other comment

**Reply by NilsBaumgartner1994** [reply]
> > > > I know what you mean, but it is only used to create the nonce in the SSO flow in directus. Since the custom endpoint does not render anything and only forwards with a 301, i would think that be ok.
> > > 
> > > 
> > > I see your point, how do you think about UX, user would rather stay at the application than open browser and redirect?
> > 
> > 
> > Oh I would love it to see my real native browser opening and going out of the app. I don’t really trust in app browsers, since it could be mocked. Also a mocked Browser is not allowed, but who knows. So I would prefer in terms of safety to be moved out of my app.
> 
> For "in app browser", we're not going to use in app webview, we will use their SDK to login right on app, not open any web view.
> 
> I see, thank you for your input, I will try with your way, hopefully it won't be rejected :D :D :D
> 
> Also, how about redirect user back to the app? Do you have any experience on this?

Yes, works like a charm with DeepLinking on Android and iOS and Web-App. I can share my solution with you. Will upload it later.

**Reply by TranQuangMinh** [reply]
> > Hi @NilsBaumgartner1994 , i have a question,
> > MyApp Android open directus to login SSO and then how to re-open app via deeplink and send directus access token or refresh token to app ?
> 
> Oh that’s easy :-) you create a custom endpoint in Directus with a „redirectTo“ param. Since the SSO will redirect you to your given application Directus sends you back to your app. Have an example
> 
> Endpoint to redirect with access token: Directus/curstomEndpoint/SSO/facebook?redirectTo=MobileApp&access_token=
> 
> Directus SSO Endpoint: Directus/login/sso?rddirectTo
> 
> In this case we set the redirect route Directus normale would set to itself, to our custom endpoint. Our custom endpoint then gets the access token and sends us back to our app. I can share a link to a gist. Meanwhile check out my added reply post at the other comment

mean is in custom endpoint, i get refresh token from cookie, and include to deeplink and redirectus to app?

**code in custom endpoint :** 

App open: ``` /auth/login/facebook?redirect=https://api-directus/custom-endpoint/custom-endpoint-after-sso ```
```js
router.get('/custom-endpoint-after-sso', async (_req: any, res: any) => {
    const deeplink = "myapp://?refresh_token=" + _req.cookies.directus_refresh_token;
    return res.redirect(deeplink)
})
```

**Reply by NilsBaumgartner1994** [reply]
> > > Hi @NilsBaumgartner1994 , i have a question,
> > > MyApp Android open directus to login SSO and then how to re-open app via deeplink and send directus access token or refresh token to app ?
> > 
> > 
> > Oh that’s easy :-) you create a custom endpoint in Directus with a „redirectTo“ param. Since the SSO will redirect you to your given application Directus sends you back to your app. Have an example
> > Endpoint to redirect with access token: Directus/curstomEndpoint/SSO/facebook?redirectTo=MobileApp&access_token=
> > Directus SSO Endpoint: Directus/login/sso?rddirectTo
> > In this case we set the redirect route Directus normale would set to itself, to our custom endpoint. Our custom endpoint then gets the access token and sends us back to our app. I can share a link to a gist. Meanwhile check out my added reply post at the other comment
> 
> mean is in custom endpoint, i get refresh token from cookie, and include to deeplink and redirectus to app?
> 
> **code in custom endpoint :**
> 
> App open: `/auth/login/facebook?redirect=https://api-directus/custom-endpoint/custom-endpoint-after-sso`
> 
> ```js
> router.get('/custom-endpoint-after-sso', async (_req: any, res: any) => {
>     const deeplink = "myapp://?refresh_token=" + _req.cookies.directus_refresh_token;
>     return res.redirect(deeplink)
> })
> ```

Correct, that is the idea behind it. You will have to get the token. Then redirect using deeplink. 
What you definitely should DO: Is to check in Directus if the url to redirect is allowed. Someone could otherwise redirect the user to their website and phish the user. Or worse, someone could use your deeplink in their app.

**Reply by TranQuangMinh** [reply]
> > > > Hi @NilsBaumgartner1994 , i have a question,
> > > > MyApp Android open directus to login SSO and then how to re-open app via deeplink and send directus access token or refresh token to app ?
> > > 
> > > 
> > > Oh that’s easy :-) you create a custom endpoint in Directus with a „redirectTo“ param. Since the SSO will redirect you to your given application Directus sends you back to your app. Have an example
> > > Endpoint to redirect with access token: Directus/curstomEndpoint/SSO/facebook?redirectTo=MobileApp&access_token=
> > > Directus SSO Endpoint: Directus/login/sso?rddirectTo
> > > In this case we set the redirect route Directus normale would set to itself, to our custom endpoint. Our custom endpoint then gets the access token and sends us back to our app. I can share a link to a gist. Meanwhile check out my added reply post at the other comment
> > 
> > 
> > mean is in custom endpoint, i get refresh token from cookie, and include to deeplink and redirectus to app?
> > **code in custom endpoint :**
> > App open: `/auth/login/facebook?redirect=https://api-directus/custom-endpoint/custom-endpoint-after-sso`
> > ```js
> > router.get('/custom-endpoint-after-sso', async (_req: any, res: any) => {
> >     const deeplink = "myapp://?refresh_token=" + _req.cookies.directus_refresh_token;
> >     return res.redirect(deeplink)
> > })
> > ```
> 
> Correct, that is the idea behind it. You will have to get the token. Then redirect using deeplink. What you definitely should DO: Is to check in Directus if the url to redirect is allowed. Someone could otherwise redirect the user to their website and phish the user. Or worse, someone could use your deeplink in their app.

Many thank @NilsBaumgartner1994. :D

**Reply by aidenfoxx** [reply]
Thanks for fielding these questions @NilsBaumgartner1994!

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
